File: facade/src/main/java/org/tdf/sunflower/facade/AbstractConsensusEngine.java
package org.tdf.sunflower.facade;

import lombok.Getter;
import lombok.Setter;
import org.tdf.common.event.EventBus;
import org.tdf.common.store.Store;
import org.tdf.common.trie.Trie;
import org.tdf.common.util.HexBytes;
import org.tdf.sunflower.state.Account;
import org.tdf.sunflower.state.BuiltinContract;
import org.tdf.sunflower.state.StateTrie;
import org.tdf.sunflower.types.Block;
import org.tdf.sunflower.types.ConsensusConfig;

import java.util.Collections;
import java.util.List;


@Getter
@Setter
public abstract class AbstractConsensusEngine implements ConsensusEngine {
    public static final AbstractConsensusEngine NONE = new AbstractConsensusEngine() {
        @Override
        public void init(ConsensusConfig config) {
            if (getTransactionPool() == null)
                throw new RuntimeException("transaction pool not injected");
            if (getSunflowerRepository() == null)
                throw new RuntimeException("consortium repository not injected");
            if (getEventBus() == null) {
                throw new RuntimeException("event bus not injected");
            }

            setMiner(Miner.NONE);
            setValidator(Validator.NONE);
            setGenesisBlock(new Block());
            setPeerServerListener(PeerServerListener.NONE);
        }

        @Override
        public String getName() {
            return "none";
        }
    };

    // contract storage trie
    private Trie<HexBytes, HexBytes> contractStorageTrie;
    // a map between hash and wasm byte code
    private Store<HexBytes, HexBytes> contractCodeStore;
    // sub class should set miner explicitly when init() called
    private Miner miner;
    // sub class should set validator explicitly when init() called
    private Validator validator;
    // account stateTrie will be injected before init() called
    private StateTrie<HexBytes, Account> accountTrie;
    // sub class should set genesis block explicitly when init() called
    private Block genesisBlock;
    // event bus will be injected before init() called
    private EventBus eventBus;
    // transaction pool will be injected before init() called
    private TransactionPool transactionPool;
    // sunflowerRepository will be injected before init() called
    private IRepositoryService sunflowerRepository;

    // sub class should set peer server listener explicitly when init() called
    private PeerServerListener peerServerListener = PeerServerListener.NONE;


    public AbstractConsensusEngine() {
    }

    public List<Account> getAlloc() {
        return Collections.emptyList();
    }

    public List<BuiltinContract> getBuiltins() {
        return Collections.emptyList();
    }

    public List<BuiltinContract> getBios() {
        return Collections.emptyList();
    }
}

File: facade/src/main/java/org/tdf/sunflower/facade/AbstractConsensusEngine.kt
package org.tdf.sunflower.facade

import org.tdf.common.event.EventBus
import org.tdf.common.store.Store
import org.tdf.common.trie.Trie
import org.tdf.common.util.HexBytes
import org.tdf.sunflower.state.Account
import org.tdf.sunflower.state.Builtin
import org.tdf.sunflower.state.StateTrie
import org.tdf.sunflower.types.Block
import org.tdf.sunflower.types.ConsensusConfig
import org.tdf.sunflower.types.HeaderImpl
import org.tdf.sunflower.util.FileUtils

abstract class AbstractConsensusEngine : ConsensusEngine {
    // contract storage trie, injected before init() called
    lateinit var contractStorageTrie: Trie<HexBytes, HexBytes>

    // a map between hash and wasm byte code, injected before init() called
    lateinit var contractCodeStore: Store<HexBytes, HexBytes>

    // sub class should set miner explicitly when init() called
    override var miner: Miner = Miner.NONE

    // sub class should set validator explicitly when init() called
    override var validator: Validator = Validator.NONE

    // account stateTrie will be injected before init() called
    lateinit var accountTrie: StateTrie<HexBytes, Account>

    // sub class should set genesis block explicitly when init() called
    override var genesisBlock: Block = Block(HeaderImpl())

    // event bus will be injected before init() called
    lateinit var eventBus: EventBus

    // transaction pool will be injected before init() called
    lateinit var transactionPool: TransactionPool

    // sunflowerRepository will be injected before init() called
    lateinit var repo: RepositoryService

    // sub class should set peer server listener explicitly when init() called
    override var peerServerListener = PeerServerListener.NONE

    open val alloc: Map<HexBytes, Account>
        get() = emptyMap()

    open val builtins: List<Builtin>
        get() = emptyList()
    open val bios: List<Builtin>
        get() = emptyList()

    open val code: Map<HexBytes, HexBytes> get() = emptyMap()

    companion object {
        @JvmField
        val NONE: AbstractConsensusEngine = object : AbstractConsensusEngine() {
            override fun init(config: ConsensusConfig) {
            }

            override val name: String
                get() {
                    return "none"
                }
        }
    }
}